### PR TITLE
Blockchain Data Storage Strategies

### LINKED ISSUE #6491: Article: Blockchain data storage strategies
.

**Tutorial title**

Blockchain data storage strategies

**Tutorial description**

There are several ways in which data from external sources can be stored using the Blockchain. Here they are in order of increasing cost.

1. Calldata
2. Contract code
3. Events
4. EVM storage

This article will compare the different strategies, their costs and tradeoffs, and demonstrate the necessary code to use each one, as well as the requirements to use it safely.

**Tag your tutorial**

Storage, Data availability

**Publication**

Not written yet

**Skill level**

Beginner

**Hosted on Ethereum.org or hosted elsewhere?**

I hope to host it on ethereum.org when written.

### PR DESCRIPTION
**Pull request by qbzzt**
Article that explains the strategies for data storage.

## Description

There are several ways in which data from can be stored using the Blockchain. This article will compare the different strategies, their costs and tradeoffs, as well as the requirements to use it safely.

## Related Issue

- Closes #6491

## Preview URL
- https://deploy-preview-12899--ethereumorg.netlify.app/en/developers/docs/data-availability/blockchain-data-storage-strategies/

## Summary by CodeRabbit

- **Documentation**
	- Added a new tutorial on "Blockchain Data Storage Strategies" detailing various methods for storing data on the blockchain, including costs, tradeoffs, and safety considerations.

### CODE REVIEW

**Comment by minimalsm** [comment]
@coderabbitai review

**Comment by wackerow** [comment]
Thanks @qbzzt! Will try to give this a read over the next sprint

**Comment by qbzzt** [comment]
> Thanks @qbzzt! Will try to give this a read over the next sprint

Thank you. Now I'll start working on IPFS (https://github.com/ethereum/ethereum-org-website/issues/5746)

**Comment by qbzzt** [comment]
> I've noticed the change from `fatalError("bah!")` to `return` in `discard_checking.swift`. This adjustment seems to address the error related to consuming 'self' before exiting methods that discard self. It's a nuanced fix that enhances the code's robustness. Could you provide more context on the impact of this change on pruned liveness and discard checking?

I think you posted in the wrong PR, this article does not contain any code, swift or slow.

**Comment by wackerow** [comment]
Hey @qbzzt! Taking a look through, this looks great. I do wonder though if this is appropriate for the "tutorials" section of the site. Great informational article, but doesn't really guide users through any series of actions like a tutorial usually would.

This is perfectly fine imo, I just think we should try to find another place on the site for this. We have a started page on ["Data Availability" in the dev docs](https://ethereum.org/en/developers/docs/data-availability/); I'd suggest finding a way to use this as a continuation of that content.

What do you think if we added this as a sub-page to the previous link ("Data Availability") with your proposed title ("Blockchain Data Storage Strategies")?

cc: @minimalsm

**Comment by qbzzt** [comment]
> What do you think if we added this as a sub-page to the previous link ("Data Availability") with your proposed title ("Blockchain Data Storage Strategies")?

That makes a lot more sense, you're right. Do it.

**Comment by qbzzt** [comment]
Do you want me to modify it to work as a sub-page of Data Availability? Or
do you want to do that?

**Comment by wackerow** [comment]
Okay great @qbzzt , I can make that modification

**Comment by wackerow** [comment]
Hey @qbzzt, I think this looks pretty decent relocated to the dev docs. Would you mind adding in a "Prerequisite" section just after the intro (see other dev docs pages for examples of how this usually looks). Also feel free to adjust the intro as you see fit with it's new location.

**Comment by qbzzt** [comment]
> Hey @qbzzt, I think this looks pretty decent relocated to the dev docs. Would you mind adding in a "Prerequisite" section just after the intro (see other dev docs pages for examples of how this usually looks). Also feel free to adjust the intro as you see fit with it's new location.

Done.

File: public/content/developers/docs/data-availability/blockchain-data-storage-strategies/index.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+## The C.I.A. of information security {#cia-info-security}
+
+There are three attributes we'd like to have for data.
+
+- _Confidentiality_, unauthorized entities are not allowed to read the information. This is useful in many cases, but _there are no secrets on the blockchain_. Blockchains work because everybody can verify the state transitions, so it is impossible to use them to store secrets directly. There are ways to store confidential information on the blockchain, but they all rely on some offchain component to store at least a key.
+
+- _Integrity_, the information is correct, it cannot be changed by unauthorized entities, or in unauthorized ways (for example, transferring [ERC-20 tokens](https://eips.ethereum.org/EIPS/eip-20#events) without a `Transfer` event). On the blockchain, every node verifies every state change, which ensures integrity.
+
+- _Availability_, the information is available to any authorized entity. On the blockchain, this is usually achieved by having the information available on every [full node](https://ethereum.org/developers/docs/nodes-and-clients#full-node).
```
Comment (wackerow) [inline public/content/developers/docs/data-availability/blockchain-data-storage-strategies/index.md:35]: @qbzzt I like this section, but it feels awkwardly placed... how does this place into the rest of the content? Maybe we add a little more "intro" text to tell the user what to expect on this page, and maybe set the stage for why we're showing the CIA section here.

Comment (qbzzt) [reply public/content/developers/docs/data-availability/blockchain-data-storage-strategies/index.md]: I moved and modified, and I think it makes more sense now.

**Review by wackerow — APPROVED** [review]
Looks good! Thanks @qbzzt!

**Pull request merged**